rfcs/2021-06-15-shipping-blade.md

```diff
@@ -460,6 +466,178 @@ One common question you might have: If exports is not supported by all the tools
+
+So what do we do?
+
+We need to run a script that does this for us. In simpler terms we need to copy these files ourselves. We'll use [`copyfiles`](https://www.npmjs.com/package/copyfiles) for this which will copy our custom `*.d.ts` files to the exact location under `build/types` as in source. Following is the script that we need to run
+```bash
+copyfiles -u 1 \"src/**/*.d.ts\" build/types
+```
+So here are the steps in a nutshell we need to perform to emit types and copy our custom `.d.ts` files:
+1. Run `tsc` to emit types to `build/types`
+2. Run `copyfiles` to copy our custom `*.d.ts` files to `build/types`
+
+After doing the above steps here's the build directory structure that we'll get
+```
+build/
+└── types/
+    ├── components/
+    │   ├── ThemeProvider/
+    │   │   └── index.d.ts
+    │   └── index.d.ts
+    ├── tokens/
+    │   ├── theme/
+    │   │   ├── paymentTheme.d.ts
+    │   │   └── theme.d.ts
+    │   └── index.d.ts
+    └── utils/
+        ├── getColorScheme/
+        │   ├── getColorScheme.d.ts 
+        │   └── index.d.ts
+        └── index.d.ts
+```
+So we have our types emitted as well as our custom `*.d.ts` files copied under `build/types`
+
+### Phase 3
+The next step is to run rollup to bundle our source code. Once we run rollup here's how the `build` directory will look like.
+```
+build/
+├── components/
+│   ├── index.web.js
+│   └── index.native.js
+├── tokens/
+│   ├── index.web.js
+│   └── index.native.js
+├── utils/
+│   ├── index.web.js
+│   └── index.native.js  
+└── types/
+    ├── components/
+    │   ├── ThemeProvider/
+    │   │   └── index.d.ts
+    │   └── index.d.ts
+    ├── tokens/
+    │   ├── theme/
+    │   │   ├── paymentTheme.d.ts
+    │   │   └── theme.d.ts
+    │   └── index.d.ts
+    └── utils/
+        ├── getColorScheme/
+        │   ├── getColorScheme.d.ts 
+        │   └── index.d.ts
+        └── index.d.ts
+```
+Now do you see any issues?
+
+The issue is our source code got bundled into `components`, `tokens` and `utils` but we have our `types` structure same as our source code. So how will TS map types when the consumers will consume our library? It won't. 
+
+So what do we do?
+
+### Phase 4
+We need to bundle our declarations so that it maps to `components`, `tokens` and `utils`. We'll use rollup to bundle our declarations with [`rollup-plugin-dts`](https://github.com/Swatinem/rollup-plugin-dts).
+
+Here's how our rollup config will look like:
+```js
+const exportCategories = ['components', 'tokens', 'utils'];
+
+const getDeclarationsConfig = ({ exportCategory }) => ({
+  input: `${outputRootDirectory}/types/${exportCategory}/index.d.ts`,
```

Comment:
This section is not as clear. Seems like there's some magic going on in this phase 😛 

For instance, how does this step handle generating config for a nested directory like `/types/components/ThemeProvider/index.d.ts`

`exportCategory` will map it till `/types/components`. How does it figure out that there's another directory inside that it needs to go to and pick up that `index.d.ts` as well?

Reply:
it won't. We are just using top level categories' re-exported file. Look at the directory structure above. we have

```
src/
	components/
		index.d.ts --- this has all the re-exported things and points to nested things and stuff
	tokens/
		index.d.ts --- this has all the re-exported things and points to nested things and stuff
	utils/
		index.d.ts --- this has all the re-exported things and points to nested things and stuff
```

it's same as giving entry point while bundling. We just define one entry point in any of the bundling tool and then it goes and generates dependency graph. Not sure if it makes sense to add this much detail in RFC?

Reply:
Got it. The plugin is internally doing this but unless a reader knows this, it seems like there's a bug here. Maybe mention `the plugin will pick up all the nested declaration files as well`?